GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs

```diff
@@ -0,0 +1,466 @@
+            {
+                this.tracer.RelatedError($"{nameof(this.TryQueryNewestVersion)} failed with: {ex.Message}");
+                message = ex.Message;
+                newVersion = null;
+            }
+
+            return false;
+        }
+
+        public bool TryDownloadNewestVersion(out string errorMessage)
+        {
+            if (this.latestVersion == null)
+            {
+                errorMessage = "No new version to download. Query for latest version to ensure a new version is available before downloading.";
+                return false;
+            }
+
+            using (ITracer activity = this.tracer.StartActivity(nameof(this.TryDownloadNewestVersion), EventLevel.Informational))
+            {
+                try
+                {
+                    this.DownloadedPackagePath = this.nuGetFeed.DownloadPackage(this.latestVersion.Identity).GetAwaiter().GetResult();
+                }
+                catch (Exception ex)
+                {
+                    activity.RelatedError($"{nameof(this.TryDownloadNewestVersion)} - error encountered: ${ex.Message}");
+                    errorMessage = ex.Message;
+                    return false;
+                }
+            }
+
+            errorMessage = null;
+            return true;
+        }
+
+        public bool TryCleanup(out string error)
+        {
+            error = null;
+            Exception e;
+            bool success = this.fileSystem.TryDeleteDirectory(this.localUpgradeServices.TempPath, out e);
+
+            if (!success)
+            {
+                error = e?.Message;
+                this.tracer.RelatedError($"{nameof(this.TryCleanup)} - Error encountered: {error}");
+            }
+
+            return success;
+        }
+
+        public bool TryRunInstaller(InstallActionWrapper installActionWrapper, out string error)
+        {
+            string localError = null;
+            int installerExitCode;
+            bool installSuccessful = true;
+            using (ITracer activity = this.tracer.StartActivity(nameof(this.TryRunInstaller), EventLevel.Informational))
+            {
+                try
+                {
+                    string platformKey = ReleaseManifest.WindowsPlatformKey;
+                    string upgradesDirectoryPath = ProductUpgraderInfo.GetUpgradesDirectoryPath();
+
+                    Exception e;
+                    if (!this.fileSystem.TryDeleteDirectory(this.localUpgradeServices.TempPath, out e))
+                    {
+                        error = e?.Message;
+                        return false;
+                    }
+
+                    string extractedPackagePath = this.UnzipPackageToTempLocation();
+                    this.releaseManifest = ReleaseManifest.FromJsonFile(Path.Combine(extractedPackagePath, "content", "install-manifest.json"));
+                    InstallManifestPlatform platformInstallManifest = this.releaseManifest.PlatformInstallManifests[platformKey];
```

Comment:
Use `TryGetValue` instead of `[platformKey]`? 

It looks like `Dictionary<TKey,TValue>.Item[TKey]` will throw a `KeyNotFoundException` if the key does not exist (https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.item?view=netframework-4.6.2) and so we won't ever hit the `if (platformInstallManifest == null)` case.